### ISSUE TITLE
Tab key is extremely slow on medium-large files when emmet.triggerExpansionOnTab is enabled

### ISSUE THREAD

**Issue by xak2000** [microsoft/vscode#71996]
Issue Type: <b>Bug</b>

When I open a large file (tried on `.php` and `.html`) and `emmet.triggerExpansionOnTab` option is enabled, then each pressing on <kbd>TAB</kbd> key (even on blank line) make my computer hung for about 1-2 seconds (depends on file size) with high CPU load.

This is true with current 1.33 version of VS CODE, with or without extensions (tried with `code --disable-extensions`), with or without project open. I also tried to download 1.34 insiders build and the behavior is still the same.

Steps to Reproduce:

1. Download a fresh 1.34 insiders build (or use 1.33 current).
2. Create new HTML file `tmp.html` with any content. Make sure it is 2000+ lines. Better 5000+ - the hang is more obvious when the file is more large. I attached an [example file](https://github.com/Microsoft/vscode/files/3059147/tmp.html.zip) here.
3. Set cursor on any line and press <kbd>TAB</kbd>. You can press it many times in a row or press and hold - all should be fine.
4. Open `preferences` and enable `Emmet: Trigger Expansion on Tab` option (`emmet.triggerExpansionOnTab`).
5. Return to opened file and try to press <kbd>TAB</kbd> again. You should see a very notable delay between pressing the key and moving the cursor (you can enable `Toggle render whitespace` option to see this better).

    If you press and hold the <kbd>TAB</kbd> key for about 5 seconds, then you should see no spaces/tabs characters entered (but notice high CPU load), you then can move cursor to anoter place in the editor, and after another 10-30 seconds have been passed (all that time with high CPU load), the delayed space characters will appear at the current cursor position (not where the cursor been when you actually pressed the <kbd>TAB</kbd> key).

So here is two problems:

1. A high CPU load and delay between pressing <kbd>TAB</kbd> key and appearing of spaces/tab characters.
2. The space/tab characters appears after the delay not where the cursor was when <kbd>TAB</kbd> key was pressed, but at the current position of the cursor.

I notice this problem only recently. I'm not sure with which build this comes in (not used VS Code recently), but 1-2 month ago there were no this problem.

--------------

VS Code version: Code - Insiders 1.34.0-insider (56f1b4795fed05677eb82a21dac4fc1ab62d747b, 2019-04-09T05:19:45.294Z)
OS version: Windows_NT x64 6.1.7601

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-4670 CPU @ 3.40GHz (4 x 3400)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>rasterization: enabled<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|23.93GB (1.39GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>Extensions: none

**Comment by dslpp056193** [comment]
I have the same problem with latest vscode.

**Comment by octref** [comment]
Duplicate of #70371. Try tomorrow's [Insiders](https://code.visualstudio.com/insiders/) which should have the fix.

**Comment by xak2000** [comment]
@octref Tried today's insiders build. Nothing is changed.

I opened a html file with 3700 LOC, enabled `emmet.triggerExpansionOnTab`, and tried to press <key>TAB</key>. It extremely slow and consumes high CPU. For doing nothing (nothing to expand by emmet because I press <key>TAB</key> key on **empty** lines...).

Also, I can be not a duplicate of #70371 because in #70371 there is no high CPU load and there users try real emmet expansions. This bug on other hand not about emmet expansion, but about pressing <key>TAB</key> key on empty line (where emmet should not trigger at all).

VS Code version: Code - Insiders 1.37.0-insider (cf03ea3729dc37e439d724267ad29dc5840f903a, 2019-07-30T07:15:22.972Z)
OS version: Windows_NT x64 6.1.7601

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-4670 CPU @ 3.40GHz (4 x 3400)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_deferred_display_list: disabled_off<br>skia_renderer: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|23.93GB (2.88GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>Extensions: none

**Comment by octref** [comment]
It's not in insider yet. Wait a bit.

**Comment by dslpp056193** [comment]
> It's not in insider yet. Wait a bit.

Is a bit - you mean we have no chance to resolve this critycal for developer problem?)
This problem follows from when vscode released emmet from the box. We have no problem before, but now we have a lot of problem with default emmet - it does not support a lot of shortcuts from original emmet and have a lot of bugs by default installation.

I loved vscode and developed several extensions, but now i start thinking about sublime text 3 again.

**Comment by octref** [comment]
> we have no chance to resolve

It's already resolved.

> Is a bit

You can either build from source or wait for another release of Insiders version coming out this afternoon.

**Comment by xak2000** [comment]
@octref 

I downloaded a fresh copy of VS Code Insiders (I inlcuded version and commit hash on bottom of this message) and still nothing changed.

As I see, the commit 99d5ff9c0f5a66a9e1cd7d49f9af6892e13f96a4, that fixes #70371 is from PR #78207 and it is already included in insiders build (at least it is below of the current commit of downloaded build on master branch. The Insiders build is built from master, right?).

So, if you meant that this commit should fix this issue, it doesn't, unfortunately. :-(

I recored a video to show what it looks like: [vc-code-slow-tab-and-high-cpu-usage2.zip](https://github.com/microsoft/vscode/files/3451932/vc-code-slow-tab-and-high-cpu-usage2.zip). On 00:18 I pressed <kbd>TAB</kbd> key several times in a row and only on 00:43 the editor finally rendered all these spaces (in my case 4 spaces instead of "tab" character). Throughout its 25 seconds VS Code consumed 100% of a single CPU core, trying to render 10 <kbd>TAB</kbd>s (I know, that it actually consumed the cpu for something else (Emmet I think), it's just a simplification).

---

VS Code version: Code - Insiders 1.37.0-insider (d5900c642b359fe7a9eb1e3f9e2d0977f6e0905c, 2019-07-31T05:24:41.973Z)
OS version: Windows_NT x64 6.1.7601

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-4670 CPU @ 3.40GHz (4 x 3400)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: 

enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_deferred_display_list: disabled_off<br>skia_renderer: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|23.93GB (4.55GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>Extensions: none

**Comment by octref** [comment]
I'm on d5900c642b359fe7a9eb1e3f9e2d0977f6e0905c and here's what I see in a 5k line HTML file:

![fast](https://user-images.githubusercontent.com/4033249/62228498-9e00b180-b372-11e9-823a-bc5b97e80894.gif)

Whereas this is stable:

![slow](https://user-images.githubusercontent.com/4033249/62228578-cc7e8c80-b372-11e9-826f-7943402b20c4.gif)

NVM, I just realized that your bug is you are having "emmet.triggerExpansionOnTab" and expanding on empty lines, whereas the bug I fixed was for `"emmet.triggerExpansionOnTab": false`.

**Comment by DerrikMilligan** [comment]
This is indeed a bug, thanks for narrowing it down to the setting for me to turn off to at least live while it's being worked on.

**Comment by octref** [comment]
I made a cheap improvement, which is to disable expansion if the previous character is space or tab. For example `div|` expands, but `div |` doesn't. This makes tabbing on empty lines not that miserable, but a proper fix is still needed for the expansion.

**Comment by octref** [comment]
We were parsing the whole document when we didn't have to.

Emmet 's tab-to-expand would need to analyze each Node when you have multiple selections. This is to expand things like:

```
div|
div|
div|
```

with multi cursor.

However, 99% of the time tab expansion happens without any selection. I avoided parsing (which is the bottleneck that could take between 2-4s for a 5k line HTML file) for when there's no selection or when selection doesn't need the parsed document, and it's much faster now.

**Comment by andremacola** [comment]
Emmet using remote-ssh is extremely slow.

**Comment by roblourens** [comment]
@octref I see a delay of 2-3 seconds after pressing tab after `div|` in a 5000 line html file, is that still what I expect to see?

**Comment by octref** [comment]
That's not. Let me try to repro...

**Comment by octref** [comment]
I'll switch to our html languageservice parser for December.

**Comment by rzhao271** [comment]
@xak2000 @andremacola @medzhidov does this issue occur for you in the latest vscode?
I know this issue hasn't been looked at for a year, and I want to see where things are at, now.
I noticed that pressing and holding down tab is causing a bunch of events to fire in the profile, though I haven't associated any directly with Emmet yet.
Will take a closer look tomorrow.

![emmet-perf-delay](https://user-images.githubusercontent.com/7199958/101708151-8abf2500-3a41-11eb-8bc9-751db4e7a50c.PNG)

**Comment by medzhidov** [comment]
Don't worry, i started use PHPStorm, so this bug can be leaved for another one year =)

**Comment by xak2000** [comment]
@rzhao271 I checked out the file `tmp.html` that I included in the issue description and no, it doesn't occur anymore if I press and hold <kbd>TAB</kbd> on an empty line.

But I managed to reproduce the issue in other way.

1. Open preferences and enable `Emmet: Trigger Expansion on Tab` option (`emmet.triggerExpansionOnTab`).
2. Put the cursor at the start of an empty line of `tmp.html` file (e.g. on line 11).
3. Enter `div` and press <kbd>Esc</kbd> to close the autocompletion popup menu.
4. Press and hold <kbd>TAB</kbd> for 5-10 seconds. Release it.
5. Nothing new will be entered in the editor (emmet expansion would not happen) in next 10-20 seconds (one expansion sometimes still happens so one instance of `<tab></tab>` text could appear), but you will notice one CPU thread that consume 100% CPU. Main editor thread is still responsive, I suppose. I can move cursor to another line and enter some new text here etc. It looks like VS Code is not frozen at all. To notice then something wrong I need to look at my CPU temperature and frequency and I see that it doing some hard work. And then after 10-20 seconds suddenly emmet will finish it's expansion and something like this will be printed in the editor at the location of cursor when you held TAB:

```
<div>|</div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>v></div>
```

(the `|` indicates the cursor location).

And what is worse: cursor will be moved back into the position of the file when you pressed <kbd>TAB</kbd> (after first `<div>`). Even if you already was in another place of the file and entering some text at the moment, your input will be interrupted and part of text that you entered will be placed into the new position of cursor (after first `<div>`).
